lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -254,26 +273,127 @@ instance ((PersistPublicKey (key 'AccountK)), Typeable chain)
+      where
+        ccF = build $ toText MultisigScript
         xpubF = hexF $ serializeXPub acct
         acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
 
--- | Bring a 'AccountingStyle' type back to the term-level. This requires a type
+deriving instance Ord KeyHash
+
+lookupKeyHash
+    :: (SoftDerivation k, WalletKey k)
+    => KeyHash
+    -> VerificationKeyPool k
+    -> (Maybe (Index 'Soft 'ScriptK), VerificationKeyPool k)
+lookupKeyHash keyHash pool@(VerificationKeyPool accXPub g indexedHashKeys scripts) =
+    let updateKey (ix, _) = (ix, Used)
+    in case Map.lookup keyHash indexedHashKeys of
+        Nothing -> (Nothing, pool)
+        Just (ix, _) ->
+            ( Just ix
+            , unsafeVerificationKeyPool accXPub g
+                (Map.adjust updateKey keyHash indexedHashKeys) scripts)
+
+updateKnownScripts
+    :: (WalletKey k, SoftDerivation k)
+    => (Map ScriptHash [Index 'Soft 'ScriptK] -> Map ScriptHash [Index 'Soft 'ScriptK])
+    -> VerificationKeyPool k
+    -> VerificationKeyPool k
+updateKnownScripts trKnownScripts (VerificationKeyPool accXPub g indexedHashKeys knownScripts) =
+    let knownScripts' = trKnownScripts knownScripts
+    in unsafeVerificationKeyPool accXPub g indexedHashKeys knownScripts'
+
+-- | Create a new VerificationKey pool.
+newVerificationKeyPool
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> AddressPoolGap
+    -> VerificationKeyPool k
+newVerificationKeyPool accXPub addrPoolGap =
+    unsafeVerificationKeyPool accXPub addrPoolGap Map.empty Map.empty
+
+-- | Create a new VerificationKey pool.
+-- The extension to the pool is done by adding next adjacent indices,
+-- marking them as unused and their corresponding public keys.
+unsafeVerificationKeyPool
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> AddressPoolGap
+    -> Map KeyHash (Index 'Soft 'ScriptK, AddressState)
+    -> Map ScriptHash [Index 'Soft 'ScriptK]
+    -> VerificationKeyPool k
+unsafeVerificationKeyPool accXPub num@(AddressPoolGap g) vkPoolMap knownScripts = VerificationKeyPool
+    { verPoolAccountPubKey = accXPub
+    , verPoolGap = num
+    , verPoolIndexedKeys =
+            Map.union vkPoolMap vkPoolMap'
+    , verPoolKnownScripts = knownScripts
+    }
+  where
+    minIndex = fromIntegral $ toInteger $ getIndex @'Soft minBound
+    lastNotUsedIndices =
+        L.takeWhile (\(_,isUsed) -> isUsed == Unused) $
+        L.sortOn (Down . fst) $
+        Map.elems vkPoolMap
+    firstIndexToAdd = minIndex + L.length (Map.keys vkPoolMap)
+    newIndicesToAdd = fromInteger (toInteger g) - L.length lastNotUsedIndices
+    toIndex = toEnum . fromInteger . toInteger
+    vkPoolMap' =
+        Map.fromList $
+        map (\ix -> ( calculateVerificationKeyHash $
+                        deriveVerificationKey accXPub (toIndex ix)
+                    , (toIndex ix, Unused)) )
+        [firstIndexToAdd .. firstIndexToAdd + newIndicesToAdd - 1]
```

Comment:
One of the reason why we do index elements by their derivation index is that we can look them up and know from the looked up index how much the map should be extended. For instance, when looking up an address at index `i`, we can infer how much we need to extend the map from the size of the map, the gap and this index `i` from just a simple equation involving only addition and subtraction (so it is quite efficient in practice). 

If we do this extension when reconstructing the map, we loose a crucial information which is the index that was last looked up, and this function therefore has to re-traverse the map looking for the last found index. For small maps, it doesn't matter, but as soon the map reaches a few hundred or thousands elements it can become a real performance hit. In particular, since this function is called by the SQLite, it means that everytime we deserialize a pool, we have to look for the last not used indices which involves several map to list transformation and list traversals, and it's unfortunately all for nothing because in principle we only serialize already valid and correctly extended pool to the database. So, I'll suggest the following changes to get a more efficient implementation:

-  a `extendVerificationKeyPool` function which would take an index, and extend the pool based on that index (very much like `extendAddressPool`

- Make the `unsafeVerificationKeyPool` function a mere wrapper of the constructor without any logic, to be used by external functions such as tests or the database deserialization.

Reply:
introduced `extendedVerificationKeyPool` make unsafe the `unsafeVerificationKeyPool` -> https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/b26348def673672c028919f8b2b4d5cca429c3dd